src/client/Microsoft.Identity.Client/Utils/Base64UrlHelpers.cs

```diff
@@ -2,79 +2,267 @@
+        /// <exception cref="ArgumentOutOfRangeException">offset or length is negative OR offset plus length is greater than the length of inArray.</exception>
+        public static string Encode(byte[] inArray)
+        {
+            if (inArray == null)
+                return null;
+
+            return Encode(inArray, 0, inArray.Length);
         }
 
-        public static string DecodeToString(string arg)
+        internal static string EncodeString(string str)
         {
-            byte[] decoded = DecodeToBytes(arg);
-            return CoreHelpers.CreateString(decoded);
+            if (str == null)
+                return null;
+
+            return Encode(Encoding.UTF8.GetBytes(str));
         }
 
-        public static byte[] DecodeToBytes(string arg)
+        /// <summary>
+        ///  Converts the specified string, which encodes binary data as base-64-url digits, to an equivalent 8-bit unsigned integer array.</summary>
+        /// <param name="str">base64Url encoded string.</param>
+        /// <returns>UTF8 bytes.</returns>
+        public static byte[] DecodeBytes(string str)
         {
-            string s = arg;
-            s = s.Replace(Base64UrlCharacter62, Base64Character62); // 62nd char of encoding
-            s = s.Replace(Base64UrlCharacter63, Base64Character63); // 63rd char of encoding
+            if (str == null)
+                return null;
+#if NET45
+            // 62nd char of encoding
+            str = str.Replace(base64UrlCharacter62, base64Character62);
 
-            switch (s.Length%4)
+            // 63rd char of encoding
+            str = str.Replace(base64UrlCharacter63, base64Character63);
+
+            // check for padding
+            switch (str.Length % 4)
             {
-                // Pad
                 case 0:
-                    break; // No pad chars in this case
+                    // No pad chars in this case
+                    break;
                 case 2:
-                    s += s_doubleBase64PadCharacter;
-                    break; // Two pad chars
+                    // Two pad chars
+                    str += doubleBase64PadCharacter;
+                    break;
                 case 3:
-                    s += Base64PadCharacter;
-                    break; // One pad char
+                    // One pad char
+                    str += base64PadCharacter;
+                    break;
                 default:
-                    throw new ArgumentException("Illegal base64url string!", nameof(arg));
+                    throw new FormatException($"Unable to decode: {str} as Base64url encoded string.");
             }
 
-            return Convert.FromBase64String(s); // Standard base64 decoder
+            return Convert.FromBase64String(str);
+#else
+            return UnsafeDecode(str);
+#endif
         }
 
-        internal static string Encode(byte[] arg)
+#if !NET45
+        private unsafe static byte[] UnsafeDecode(string str)
```

Comment:
`UnsafeDecodeToBytes` for consistency? (Since just `Decode`/`Encode` implies `ToString`).

Reply:
I'd like to keep the method names as they are now in Wilson. This will make it easier to copy code next time they make an improvement.